### PR TITLE
8329605: hs errfile generic events - move memory protections and nmethod flushes to separate sections

### PR DESCRIPTION
**Pull request by MBaesken**
Currently the 'generic' hs_errfile Events message log (filled  by Events::log) is sometimes rather full by messages for memory protection operations and nmethod flushes. Those seem to occur quite often and potentially move out other less frequent events, because the number of entries in the log is limited.
It might be better to separate the events into separate sections. 

The mentioned memory protection operations related entries look like this :
Event: 0.178 Protecting memory [0x000000016ebf0000,0x000000016ebfc000] with protection modes 0

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8329605](https://bugs.openjdk.org/browse/JDK-8329605): hs errfile generic events - move memory protections and nmethod flushes to separate sections (**Enhancement** - P4)

### Reviewers
 * [Lutz Schmidt](https://openjdk.org/census#lucy) (@RealLucy - **Reviewer**) ⚠️ Review applies to [7de2f9e5](https://git.openjdk.org/jdk/pull/18626/files/7de2f9e5d96834937509cd782b00a2138a234274)
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**) ⚠️ Review applies to [912afc31](https://git.openjdk.org/jdk/pull/18626/files/912afc3182497caa9ff97f24687762f02c49286b)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18626/head:pull/18626` \
`$ git checkout pull/18626`

Update a local copy of the PR: \
`$ git checkout pull/18626` \
`$ git pull https://git.openjdk.org/jdk.git pull/18626/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18626`

View PR using the GUI difftool: \
`$ git pr show -t 18626`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18626.diff">https://git.openjdk.org/jdk/pull/18626.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18626#issuecomment-2037089627)

### CODE REVIEW

**Comment by stefank** [comment]
We still have flooding in the frequent events. Instead of creating a common section for these vent, did you consider creating two new separate, specific sections for memory protection and nmethod flushing?

**Comment by MBaesken** [comment]
> We still have flooding in the frequent events. Instead of creating a common section for these vent, did you consider creating two new separate, specific sections for memory protection and nmethod flushing?

I thought about creating new specific sections. Regarding those 2,  maybe they are a bit too specific ?  But on the other hand, if others like this idea, I am fine with  it (creating sections for memory protection operations and for nmethod flushing).

**Comment by MBaesken** [comment]
> But on the other hand, if others like this idea, I am fine with it (creating sections for memory protection operations and for nmethod flushing).

I asked around in my team and added a separate section for the nmethod flush operations.

File: src/hotspot/share/utilities/events.cpp [hunk 1]
```diff
@@ -97,6 +99,8 @@ void Events::print() {
 void Events::init() {
   if (LogEvents) {
     _messages = new StringEventLog("Events", "events");
+    _nmethod_flush_messages = new StringEventLog("Nmethod flushs", "nmethodflushs");
```
Comment (TheRealMDoerr) [inline src/hotspot/share/utilities/events.cpp:102]: Should be "flushes" or "flushing events".

**Review by RealLucy — APPROVED** [review]
LGTM - given the scope of this PR.
In general, I don't like the event log to be split into multiple streams being printed separately. Yes, separate sections prevent displacement of events by other, too verbose, events. On the other hand, time coherence is lost or has to be manually re-established by the support engineer. Often enough, an issue can only be understood when seeing multiple/all events in timely order. 

Merging the event sections at print time by timestamp would be a helpful enhancement.

**Comment by stefank** [comment]
> LGTM - given the scope of this PR. In general, I don't like the event log to be split into multiple streams being printed separately. Yes, separate sections prevent displacement of events by other, too verbose, events. On the other hand, time coherence is lost or has to be manually re-established by the support engineer. Often enough, an issue can only be understood when seeing multiple/all events in timely order.
> 
> Merging the event sections at print time by timestamp would be a helpful enhancement.

FWIW, I think I am of the opposite opinion. I find it very helpful to have the events separated into distinct sections and wouldn't want them all combined into one big section.

**Comment by MBaesken** [comment]
Hi Lucy, thanks for the review !  May I have a second one please ?

> FWIW, I think I am of the opposite opinion. I find it very helpful to have the events separated into distinct sections and wouldn't 
> want them all combined into one big section.

Yeah, both approaches (one single big log, or multiple ones) have pros and cons.
But having multiple logs is a long established HS approach so I do not think that it is in scope of this PR to change the established approach.

**Comment by RealLucy** [comment]
> ... I do not think that it is in scope of this PR ...
Oh no, I didn't want to suggest to do such a change in the scope of this PR. And yes, I agree, there are reasons why you would want separate streams. As always, a solution that suits all needs is hard to find.

**Review by TheRealMDoerr — COMMENTED** [review]
I think it's ok if @stefank also likes it.

**Comment by tstuefe** [comment]
Hmm, I think the separation makes sense, but I don't like the "frequent" moniker. All other event logs are separated by thematic area. "frequent" is orthogonal to that. I can have frequent/non-frequent class loading messages, or exceptions. 

My proposal would be either to drop these memory protection events (do we need them? or are they remnants of some old support issues?) or to put them into a 'memprot' section or similar.

**Comment by stefank** [comment]
I agree with @tstuefe

**Comment by MBaesken** [comment]
>My proposal would be either to drop these memory protection events (do we need them? or are they remnants of some old >support issues?) or to put them into a 'memprot' section or similar.

A separate memprotect section would be good I can add this.
I think usually these protection events appear around/after threads are added.  I think we had some issues with these mem protections in the past so I would be cautious to completely removed them.

**Review by stefank — APPROVED** [review]

**Comment by MBaesken** [comment]
I added a memprotect section.

**Comment by MBaesken** [comment]
Hi Stefan, thanks for the review !

**Review by tstuefe — APPROVED** [review]
Good, thank you

**Pull request closed without merging**